lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart

```diff
@@ -0,0 +1,132 @@
+    filterQuality.index
+  )));
+
+  factory SkwasmImageFilter.fromColorFilter(SkwasmColorFilter filter) =>
+    SkwasmImageFilter._(imageFilterCreateFromColorFilter(filter.handle));
+
+  factory SkwasmImageFilter.fromUiFilter(ui.ImageFilter filter) {
+    if (filter is ui.ColorFilter) {
+      final SkwasmColorFilter colorFilter =
+        SkwasmColorFilter.fromEngineColorFilter(filter as EngineColorFilter);
+      final SkwasmImageFilter outputFilter = SkwasmImageFilter.fromColorFilter(colorFilter);
+      colorFilter.dispose();
+      return outputFilter;
+    } else {
+      return filter as SkwasmImageFilter;
+    }
+  }
+
+  factory SkwasmImageFilter.compose(
+    ui.ImageFilter outer,
+    ui.ImageFilter inner,
+  ) {
+    final SkwasmImageFilter nativeOuter = SkwasmImageFilter.fromUiFilter(outer);
+    final SkwasmImageFilter nativeInner = SkwasmImageFilter.fromUiFilter(inner);
+    return SkwasmImageFilter._(imageFilterCompose(nativeOuter.handle, nativeInner.handle));
+  }
+
+  void dispose() {
+    if (!_isDisposed) {
+      imageFilterDispose(handle);
+      _isDisposed = true;
+    }
+  }
+
+  final ImageFilterHandle handle;
+  bool _isDisposed = false;
+
+  @override
+  ui.Rect filterBounds(ui.Rect inputBounds) => withStackScope((StackScope scope) {
+    final RawIRect rawRect = scope.convertIRectToNative(inputBounds);
+    imageFilterGetFilterBounds(handle, rawRect);
+    return scope.convertIRectFromNative(rawRect);
+  });
+}
+
+class SkwasmColorFilter {
+  SkwasmColorFilter._(this.handle);
+
+  factory SkwasmColorFilter.fromEngineColorFilter(EngineColorFilter colorFilter) =>
+    switch (colorFilter.type) {
+      ColorFilterType.mode => SkwasmColorFilter._(colorFilterCreateMode(
+        colorFilter.color!.value,
+        colorFilter.blendMode!.index,
+      )),
+      ColorFilterType.linearToSrgbGamma => SkwasmColorFilter._(colorFilterCreateLinearToSRGBGamma()),
+      ColorFilterType.srgbToLinearGamma => SkwasmColorFilter._(colorFilterCreateSRGBToLinearGamma()),
+      ColorFilterType.matrix => withStackScope((StackScope scope) {
+        final Pointer<Float> nativeMatrix = scope.convertDoublesToNative(colorFilter.matrix!);
+        return SkwasmColorFilter._(colorFilterCreateMatrix(nativeMatrix));
+      }),
+    };
+
+  factory SkwasmColorFilter.composed(
+    SkwasmColorFilter outer,
+    SkwasmColorFilter inner,
+  ) => SkwasmColorFilter._(colorFilterCompose(outer.handle, inner.handle));
+
+  void dispose() {
+    if (!_isDisposed) {
+      colorFilterDispose(handle);
+      _isDisposed = true;
+    }
+  }
+
+  final ColorFilterHandle handle;
+  bool _isDisposed = false;
+}
+
+class SkwasmMaskFilter {
+  SkwasmMaskFilter._(this.handle);
+
+  factory SkwasmMaskFilter.fromUiMaskFilter(ui.MaskFilter maskFilter) =>
+    SkwasmMaskFilter._(maskFilterCreateBlur(
+      maskFilter.webOnlyBlurStyle.index,
+      maskFilter.webOnlySigma
+    ));
+
+  final MaskFilterHandle handle;
+  bool _isDisposed = false;
+
+  void dispose() {
```

Comment:
I believe the pattern we use is to `assert(!_isDisposed)` in all dispose methods because double-dispose likely points to some app-side bug.

Reply:
Well, I don't actually have the finalization registry wired up yet, but the idea is for this to be safe to call more than once, since it might be called manually by the user and it might be called by the finalization registry. I think it makes sense for this to be idempotent, at least for how I envision this being used with the `UniqueRef` stuff.

Reply:
Yeah, as long as the object reference does not leak to the framework then it should be fine. The trouble starts when the framework references the object. Then you have a chicken and egg problem, the framework drops the reference and the object is GC'd, but then `FinalizationRegistry` doesn't have an object to call `dispose()` on. In CanvasKit each Skia object is represented using a pair of Dart objects, a `Ck*` object (e.g. `CkPaint`) and a `UniqueRef`. Then we do `FinalizationRegistry.register(ckObject, uniqueRef)` and hand the `ckObject` reference to the framework. When `ckObject` is GC'd, we call `uniqueRef.dispose()`. `UniqueRef` is the only place that could potentially be disposed more than once, once manually and once via `FinalizationRegistry`. However, `CkObject` should only be disposed once, which is where we put `assert(!isDisposed)`.

Reply:
I'm guessing in Skwasm we don't need `UniqueRef` at all. I think `Pointer` can play the same role without an extra wrapper.

Reply:
Yeah I see what you're saying. I know `FinalizationRegistry` basically needs an extra layer of indirection, but I haven't quite figured out how this is all going to work with the Skwasm objects. If you don't mind, I'm just going to kick the can on this issue and revisit it when I hook up the finalization registry stuff.